### DISCUSSION TITLE
Upgrading old 2.2 app to 6.3?

### DISCUSSION THREAD

**Discussion by danfoley** [symfony/symfony#52214] [Q&A]
How hard is this going to be? I'm a pretty experienced PHP dev and use several symfony components in my work apps. I just can't get the 2.2 app running locally so I have decided to install a fresh 6.3 install and port the code from the 2.2 app over. 

I just a really fast way to understand how 2.2 works and how 6.3 works so I know where I'm looking to port what from what.

I don't suppose any guide is out there that might assist me, maybe at least a good way to understand both versions and what they are doing. 

It's a very simple app so once I understand how the 2 versions are structured and how they do what they do I think I can handle it. Any resources besides the official docs that might help?

**Comment by ghost** [comment]
Symfony 2 and 6 differ very much. Symfony has evolved a lot since then. You can either rewrite your app or upgrade one version at the time.

**Comment by ahundiak** [comment]
Not having a working 2.2 app will complicate things.  I would wonder if I even knew what the app was doing without being concerned for how it was being done. And if I did know then I'd be tempted to just write a new app from scratch. You would probably waste more time copy/pasting/editing than if you just wrote new code.

However, from a very high level perspective, the design of Symfony 2 has not changed much. It is still a [request/response framework](https://symfony.com/doc/current/introduction/http_fundamentals.html) and once you understand that then you can at least start.  Of course many many implementation details have changed.  

I would build a 6.3 app then use the `make:controller` to get an idea of how things are wired up.  Then maybe pick a simple controller from you legacy app and try to reproduce it. You will immediately notice the directory structure has undergone major changes and that there is no longer an application level `bundle`. Also in 6.3 routing is usually done by controller attributes.  Your legacy app probably uses a routing file. You will have to understand the differences and decide if you want to move to attributes or stick with the routing file.

After that it really depends on what you app does.  Does it use Doctrine? If so then it should be too difficult to move all your entities and repositories over.  Who knows maybe your legacy app even has some testing code. Again implementation details have changed but Doctrine itself has been very resilient. Does your legacy app use authentication/authorization? Major implementation changes here.  Once again you can use the make commands to setup authentication and then mess around to get it to work with your legacy tables.

Oh and I forgot to mention. Any 2.2 third party bundles/libraries you happen to be using (such as FOSUserBundle) will almost certainly be useless on a 6x app.  So there is that to think about as well.  Good luck.
